llvm/lib/Analysis/DXILMetadataAnalysis.cpp

```diff
@@ -33,15 +38,54 @@ static ModuleMetadataInfo collectMetadataInfo(Module &M) {
     MMDAI.ValidatorVersion =
         VersionTuple(MajorMD->getZExtValue(), MinorMD->getZExtValue());
   }
+
+  // For all HLSL Shader functions
+  for (auto &F : M.functions()) {
+    if (!F.hasFnAttribute("hlsl.shader"))
+      continue;
+
+    EntryProperties EFP{};
+    // Get "hlsl.shader" attribute
+    Attribute EntryAttr = F.getFnAttribute("hlsl.shader");
+    assert(EntryAttr.isValid() &&
+           "Invalid value specified for HLSL function attribute hlsl.shader");
+    StringRef EntryProfile = EntryAttr.getValueAsString();
+    Triple T("", "", "", EntryProfile);
+    EFP.ShaderStage = T.getEnvironment();
+    // Get numthreads attribute value, if one exists
+    StringRef NumThreadsStr =
+        F.getFnAttribute("hlsl.numthreads").getValueAsString();
+    if (!NumThreadsStr.empty()) {
+      SmallVector<StringRef> NumThreadsVec;
+      NumThreadsStr.split(NumThreadsVec, ',');
+      assert(NumThreadsVec.size() == 3 && "Invalid numthreads specified");
+      // Read in the three component values of numthreads
+      if (!llvm::to_integer(NumThreadsVec[0], EFP.NumThreadsX, 10))
+        assert(false && "Failed to parse X component of numthreads");
+      if (!llvm::to_integer(NumThreadsVec[1], EFP.NumThreadsY, 10))
+        assert(false && "Failed to parse Y component of numthreads");
+      if (!llvm::to_integer(NumThreadsVec[2], EFP.NumThreadsZ, 10))
+        assert(false && "Failed to parse Z component of numthreads");
+    }
+    MMDAI.EntryPropertyMap.insert(std::make_pair(std::addressof(F), EFP));
+  }
   return MMDAI;
 }
 
 void ModuleMetadataInfo::print(raw_ostream &OS) const {
   OS << "Shader Model Version : " << ShaderModelVersion.getAsString() << "\n";
   OS << "DXIL Version : " << DXILVersion.getAsString() << "\n";
-  OS << "Shader Stage : " << Triple::getEnvironmentTypeName(ShaderStage)
+  OS << "Target Shader Stage : " << Triple::getEnvironmentTypeName(ShaderStage)
      << "\n";
   OS << "Validator Version : " << ValidatorVersion.getAsString() << "\n";
+  for (auto MapItem : EntryPropertyMap) {
```

Comment:
We can use C++17 so you could use structured bindings to clean this up slightly if you want. In any case this should probably be `const auto &` to avoid copying the items unnecessarily